File: yupana-jdbc/src/main/java/org/yupana/jdbc/YupanaPreparedStatement.java
/*
 * Copyright 2019 Rusexpertiza LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.yupana.jdbc;

import org.yupana.Proto;
import org.yupana.api.query.Result;

import java.io.InputStream;
import java.io.Reader;
import java.math.BigDecimal;
import java.net.URL;
import java.sql.*;
import java.util.*;
import java.util.logging.Level;
import java.util.logging.Logger;

class YupanaPreparedStatement extends YupanaStatement implements PreparedStatement {

    public YupanaPreparedStatement(YupanaConnection connection, String templateQuery) {
        super(connection);
        this.templateQuery = templateQuery;
    }

    private static Logger LOGGER = Logger.getLogger(YupanaPreparedStatement.class.getName());

    private String templateQuery;
    private Map<Integer, Proto.ParameterValue> parameters = new HashMap<>();
    private List<Map<Integer, Proto.ParameterValue>> batch = new ArrayList<>();

    private void setParameter(int idx, Proto.ParameterValue v) {
        parameters.put(idx, v);
    }

    private void setParameter(int idx, BigDecimal v) {
        setParameter(idx, Proto.ParameterValue.newBuilder()
                .setValue(Proto.Value.newBuilder()
                        .setDecimalValue(v.toString())).build());
    }

    @Override
    public void addBatch() throws SQLException {
        batch.add(parameters);
        parameters = new HashMap<>();
    }

    @Override
    public void clearParameters() throws SQLException {
        parameters = new HashMap<>();
    }

    @Override
    public void clearBatch() throws SQLException {
        batch.clear();
    }

    @Override
    public boolean execute() throws SQLException {
        LOGGER.log(Level.FINE, "Execute prepared statement {0}", templateQuery);
        Result result = connection.runQuery(templateQuery, parameters);
        lastResultSet = new YupanaResultSet(this, result);
        return true;
    }

    @Override
    public int[] executeBatch() throws SQLException {
        LOGGER.log(Level.FINE, "Execute prepared statement {0}", templateQuery);
        if (batch.isEmpty()) throw new SQLException("Batch is not defined");
        Result result = connection.runBatchQuery(templateQuery, batch);
        lastResultSet = new YupanaResultSet(this, result);
        int[] status = new int[batch.size()];
        Arrays.fill(status, 1);
        return status;
    }

    @Override
    public ResultSet executeQuery() throws SQLException {
        execute();
        return lastResultSet;
    }

    @Override
    public int executeUpdate() throws SQLException {
        return 0;
    }

    @Override
    public ResultSetMetaData getMetaData() throws SQLException {
        return null;
    }

    @Override
    public ParameterMetaData getParameterMetaData() throws SQLException {
        return null;
    }


    @Override
    public void setByte(int parameterIndex, byte x) throws SQLException {
        setParameter(parameterIndex, BigDecimal.valueOf(x));
    }

    @Override
    public void setShort(int parameterIndex, short x) throws SQLException {
        setParameter(parameterIndex,BigDecimal.valueOf(x));
    }

    @Override
    public void setInt(int parameterIndex, int x) throws SQLException {
        setParameter(parameterIndex, BigDecimal.valueOf(x));
    }

    @Override
    public void setLong(int parameterIndex, long x) throws SQLException {
        setParameter(parameterIndex, BigDecimal.valueOf(x));
    }

    @Override
    public void setFloat(int parameterIndex, float x) throws SQLException {
        setParameter(parameterIndex, BigDecimal.valueOf(x));
    }

    @Override
    public void setDouble(int parameterIndex, double x) throws SQLException {
        setParameter(parameterIndex, BigDecimal.valueOf(x));
    }

    @Override
    public void setBigDecimal(int parameterIndex, BigDecimal x) throws SQLException {
        setParameter(parameterIndex, x);
    }

    @Override
    public void setString(int parameterIndex, String x) throws SQLException {
        setParameter(parameterIndex, Proto.ParameterValue.newBuilder().setValue(Proto.Value.newBuilder().setTextValue(x)).build());
    }

    @Override
    public void setTimestamp(int parameterIndex, Timestamp x) throws SQLException {
        setParameter(parameterIndex, Proto.ParameterValue.newBuilder().setValue(Proto.Value.newBuilder().setTimeValue(x.getTime())).build());
    }

    @Override
    public void setArray(int parameterIndex, Array x) throws SQLException {
        throw new SQLFeatureNotSupportedException("SqlArrays are not supported");
    }

    @Override
    public void setAsciiStream(int parameterIndex, InputStream x, int length) throws SQLException {
        throw new SQLFeatureNotSupportedException("InputStreams are not supported");
    }

    @Override
    public void setAsciiStream(int parameterIndex, InputStream x, long length) throws SQLException {
        throw new SQLFeatureNotSupportedException("InputStreams are not supported");
    }

    @Override
    public void setAsciiStream(int parameterIndex, InputStream x) throws SQLException {
        throw new SQLFeatureNotSupportedException("InputStreams are not supported");
    }

    @Override
    public void setBinaryStream(int parameterIndex, InputStream x, int length) throws SQLException {
        throw new SQLFeatureNotSupportedException("InputStreams are not supported");
    }

    @Override
    public void setBinaryStream(int parameterIndex, InputStream x, long length) throws SQLException {
        throw new SQLFeatureNotSupportedException("InputStreams are not supported");
    }

    @Override
    public void setBinaryStream(int parameterIndex, InputStream x) throws SQLException {
        throw new SQLFeatureNotSupportedException("InputStreams are not supported");
    }

    @Override
    public void setBlob(int parameterIndex, Blob x) throws SQLException {
        throw new SQLFeatureNotSupportedException("Blobs are not supported");
    }

    @Override
    public void setBlob(int parameterIndex, InputStream inputStream, long length) throws SQLException {
        throw new SQLFeatureNotSupportedException("Blobs are not supported");
    }

    @Override
    public void setBlob(int parameterIndex, InputStream inputStream) throws SQLException {
        throw new SQLFeatureNotSupportedException("Blobs are not supported");
    }

    @Override
    public void setBoolean(int parameterIndex, boolean x) throws SQLException {
        throw new SQLFeatureNotSupportedException("Booleans are not supported");
    }

    @Override
    public void setBytes(int parameterIndex, byte[] x) throws SQLException {
        throw new SQLFeatureNotSupportedException("Bytes are not supported");
    }

    @Override
    public void setCharacterStream(int parameterIndex, Reader reader, int length) throws SQLException {
        throw new SQLFeatureNotSupportedException("Character streams are not supported");
    }

    @Override
    public void setCharacterStream(int parameterIndex, Reader reader, long length) throws SQLException {
        throw new SQLFeatureNotSupportedException("Character streams are not supported");
    }

    @Override
    public void setCharacterStream(int parameterIndex, Reader reader) throws SQLException {
        throw new SQLFeatureNotSupportedException("Character streams are not supported");
    }

    @Override
    public void setNCharacterStream(int parameterIndex, Reader value, long length) throws SQLException {
        throw new SQLFeatureNotSupportedException("Character streams are not supported");
    }

    @Override
    public void setNCharacterStream(int parameterIndex, Reader value) throws SQLException {
        throw new SQLFeatureNotSupportedException("Character streams are not supported");
    }

    @Override
    public void setUnicodeStream(int parameterIndex, InputStream x, int length) throws SQLException {
        throw new SQLFeatureNotSupportedException("InputStreams are not supported");
    }

    @Override
    public void setClob(int parameterIndex, Clob x) throws SQLException {
        throw new SQLFeatureNotSupportedException("Clobs are not supported");
    }

    @Override
    public void setNClob(int parameterIndex, NClob value) throws SQLException {
        throw new SQLFeatureNotSupportedException("Clobs are not supported");
    }

    @Override
    public void setClob(int parameterIndex, Reader reader, long length) throws SQLException {
        throw new SQLFeatureNotSupportedException("Clobs are not supported");
    }

    @Override
    public void setNClob(int parameterIndex, Reader reader, long length) throws SQLException {
        throw new SQLFeatureNotSupportedException("Clobs are not supported");
    }

    @Override
    public void setClob(int parameterIndex, Reader reader) throws SQLException {
        throw new SQLFeatureNotSupportedException("Clobs are not supported");
    }

    @Override
    public void setNClob(int parameterIndex, Reader reader) throws SQLException {
        throw new SQLFeatureNotSupportedException("Clobs are not supported");
    }

    @Override
    public void setDate(int parameterIndex, java.sql.Date x) throws SQLException {
        throw new SQLFeatureNotSupportedException("Dates are not supported");
    }

    @Override
    public void setDate(int parameterIndex, java.sql.Date x, Calendar cal) throws SQLException {
        throw new SQLFeatureNotSupportedException("Dates are not supported");
    }

    @Override
    public void setTime(int parameterIndex, Time x, Calendar cal) throws SQLException {
        throw new SQLFeatureNotSupportedException("Time is not supported");
    }

    @Override
    public void setTime(int parameterIndex, Time x) throws SQLException {
        throw new SQLFeatureNotSupportedException("Time is not supported");
    }

    @Override
    public void setTimestamp(int parameterIndex, Timestamp x, Calendar cal) throws SQLException {
        throw new SQLFeatureNotSupportedException("Calendars are not supported");
    }

    @Override
    public void setObject(int parameterIndex, Object x, int targetSqlType) throws SQLException {
        throw new SQLFeatureNotSupportedException("Objects are not supported");
    }

    @Override
    public void setObject(int parameterIndex, Object x) throws SQLException {
        throw new SQLFeatureNotSupportedException("Objects are not supported");
    }

    @Override
    public void setObject(int parameterIndex, Object x, int targetSqlType, int scaleOrLength) throws SQLException {
        throw new SQLFeatureNotSupportedException("Objects are not supported");
    }

    @Override
    public void setNull(int parameterIndex, int sqlType) throws SQLException {
        throw new SQLFeatureNotSupportedException("Nulls are not supported");
    }

    @Override
    public void setNull(int parameterIndex, int sqlType, String typeName) throws SQLException {
        throw new SQLFeatureNotSupportedException("Nulls are not supported");
    }

    @Override
    public void setRef(int parameterIndex, Ref x) throws SQLException {
        throw new SQLFeatureNotSupportedException("Refs are not supported");
    }

    @Override
    public void setURL(int parameterIndex, URL x) throws SQLException {
        throw new SQLFeatureNotSupportedException("URLs are not supported");
    }

    @Override
    public void setRowId(int parameterIndex, RowId x) throws SQLException {
        throw new SQLFeatureNotSupportedException("Row ids are not supported");
    }

    @Override
    public void setNString(int parameterIndex, String value) throws SQLException {
        throw new SQLFeatureNotSupportedException("NStrings are not supported");
    }

    @Override
    public void setSQLXML(int parameterIndex, SQLXML xmlObject) throws SQLException {
        throw new SQLFeatureNotSupportedException("SQLXMLs are not supported");
    }
}

File: yupana-jdbc/src/main/scala/org/yupana/jdbc/YupanaPreparedStatement.scala
/*
 * Copyright 2019 Rusexpertiza LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.yupana.jdbc

import org.yupana.protocol.{ NumericValue, ParameterValue, StringValue, TimestampValue }

import java.io.{ InputStream, Reader }
import java.net.URL
import java.sql.{ Array => SqlArray, _ }
import java.util.Calendar
import java.util.logging.{ Level, Logger }
import scala.collection.mutable.ArrayBuffer

object YupanaPreparedStatement {
  private val LOGGER: Logger = Logger.getLogger(classOf[YupanaPreparedStatement].getName)
}

class YupanaPreparedStatement protected[jdbc] (connection: YupanaConnection, templateQuery: String)
    extends YupanaStatement(connection)
    with PreparedStatement {

  private var parameters = Map.empty[Int, ParameterValue]
  private val batch = ArrayBuffer.empty[Map[Int, ParameterValue]]

  private def setParameter(idx: Int, v: ParameterValue): Unit = {
    parameters += idx -> v
  }

  @throws[SQLException]
  override def addBatch(): Unit = {
    batch += parameters
    parameters = Map.empty
  }

  @throws[SQLException]
  override def clearParameters(): Unit = {
    parameters = Map.empty
  }

  @throws[SQLException]
  override def clearBatch(): Unit = {
    batch.clear()
  }

  @throws[SQLException]
  override def execute: Boolean = {
    YupanaPreparedStatement.LOGGER.log(Level.FINE, "Execute prepared statement {0}", templateQuery)
    val result = connection.runQuery(templateQuery, parameters)
    lastResultSet = new YupanaResultSet(this, result)
    true
  }

  @throws[SQLException]
  override def executeBatch: Array[Int] = {
    YupanaPreparedStatement.LOGGER.log(Level.FINE, "Execute prepared statement {0}", templateQuery)
    if (batch.isEmpty) throw new SQLException("Batch is not defined")
    val result = connection.runBatchQuery(templateQuery, batch.toSeq)
    lastResultSet = new YupanaResultSet(this, result)
    Array.fill(batch.size)(1)
  }

  @throws[SQLException]
  override def executeQuery: ResultSet = {
    execute
    lastResultSet
  }

  @throws[SQLException]
  override def executeUpdate: Int = 0

  @throws[SQLException]
  override def getMetaData: ResultSetMetaData = null

  @throws[SQLException]
  override def getParameterMetaData: ParameterMetaData = null

  @throws[SQLException]
  override def setArray(parameterIndex: Int, x: SqlArray): Unit = {
    throw new SQLFeatureNotSupportedException("SqlArrays are not supported")
  }

  @throws[SQLException]
  override def setAsciiStream(parameterIndex: Int, x: InputStream): Unit = {
    throw new SQLFeatureNotSupportedException("InputStreams are not supported")
  }

  @throws[SQLException]
  override def setAsciiStream(parameterIndex: Int, x: InputStream, length: Int): Unit = {
    throw new SQLFeatureNotSupportedException("InputStreams are not supported")
  }

  @throws[SQLException]
  override def setAsciiStream(parameterIndex: Int, x: InputStream, length: Long): Unit = {
    throw new SQLFeatureNotSupportedException("InputStreams are not supported")
  }

  @throws[SQLException]
  override def setBigDecimal(parameterIndex: Int, x: java.math.BigDecimal): Unit = {
    setParameter(parameterIndex, NumericValue(x))
  }

  @throws[SQLException]
  override def setBinaryStream(parameterIndex: Int, x: InputStream): Unit = {
    throw new SQLFeatureNotSupportedException("InputStreams are not supported")
  }

  @throws[SQLException]
  override def setBinaryStream(parameterIndex: Int, x: InputStream, length: Int): Unit = {
    throw new SQLFeatureNotSupportedException("InputStreams are not supported")
  }

  @throws[SQLException]
  override def setBinaryStream(parameterIndex: Int, x: InputStream, length: Long): Unit = {
    throw new SQLFeatureNotSupportedException("InputStreams are not supported")
  }

  @throws[SQLException]
  override def setBlob(parameterIndex: Int, x: Blob): Unit = {
    throw new SQLFeatureNotSupportedException("Blobs are not supported")
  }

  @throws[SQLException]
  override def setBlob(parameterIndex: Int, x: InputStream): Unit = {
    throw new SQLFeatureNotSupportedException("Blobs are not supported")
  }

  @throws[SQLException]
  override def setBlob(parameterIndex: Int, x: InputStream, length: Long): Unit = {
    throw new SQLFeatureNotSupportedException("Blobs are not supported")
  }

  @throws[SQLException]
  override def setBoolean(parameterIndex: Int, x: Boolean): Unit = {
    throw new SQLFeatureNotSupportedException("Booleans are not supported")
  }

  @throws[SQLException]
  override def setBytes(parameterIndex: Int, x: Array[Byte]): Unit = {
    throw new SQLFeatureNotSupportedException("Bytes are not supported")
  }

  @throws[SQLException]
  override def setCharacterStream(parameterIndex: Int, x: Reader): Unit = {
    throw new SQLFeatureNotSupportedException("Character streams are not supported")
  }

  @throws[SQLException]
  override def setCharacterStream(parameterIndex: Int, x: Reader, length: Int): Unit = {
    throw new SQLFeatureNotSupportedException("Character streams are not supported")
  }

  @throws[SQLException]
  override def setCharacterStream(parameterIndex: Int, x: Reader, length: Long): Unit = {
    throw new SQLFeatureNotSupportedException("Character streams are not supported")
  }

  @throws[SQLException]
  override def setClob(parameterIndex: Int, x: Clob): Unit = {
    throw new SQLFeatureNotSupportedException("Clobs are not supported")
  }

  @throws[SQLException]
  override def setClob(parameterIndex: Int, x: Reader): Unit = {
    throw new SQLFeatureNotSupportedException("Clobs are not supported")
  }

  @throws[SQLException]
  override def setClob(parameterIndex: Int, x: Reader, length: Long): Unit = {
    throw new SQLFeatureNotSupportedException("Clobs are not supported")
  }

  @throws[SQLException]
  override def setDate(parameterIndex: Int, x: Date): Unit = {
    setParameter(parameterIndex, TimestampValue(x.getTime))
  }

  @throws[SQLException]
  override def setDate(parameterIndex: Int, x: Date, cal: Calendar): Unit = {
    throw new SQLFeatureNotSupportedException("Dates are not supported")
  }

  @throws[SQLException]
  override def setDouble(parameterIndex: Int, x: Double): Unit = {
    setParameter(parameterIndex, NumericValue(x))
  }

  @throws[SQLException]
  override def setFloat(parameterIndex: Int, x: Float): Unit = {
    setParameter(parameterIndex, NumericValue(BigDecimal(x.toDouble)))
  }

  @throws[SQLException]
  override def setInt(parameterIndex: Int, x: Int): Unit = {
    setParameter(parameterIndex, NumericValue(x))
  }

  @throws[SQLException]
  override def setByte(parameterIndex: Int, x: Byte): Unit = {
    setParameter(parameterIndex, NumericValue(x))
  }

  @throws[SQLException]
  override def setLong(parameterIndex: Int, x: Long): Unit = {
    setParameter(parameterIndex, NumericValue(x))
  }

  @throws[SQLException]
  override def setNCharacterStream(parameterIndex: Int, x: Reader): Unit = {
    throw new SQLFeatureNotSupportedException("Character streams are not supported")
  }

  @throws[SQLException]
  override def setNCharacterStream(parameterIndex: Int, x: Reader, length: Long): Unit = {
    throw new SQLFeatureNotSupportedException("Character streams are not supported")
  }

  @throws[SQLException]
  override def setNClob(parameterIndex: Int, x: NClob): Unit = {
    throw new SQLFeatureNotSupportedException("NClobs are not supported")
  }

  @throws[SQLException]
  override def setNClob(parameterIndex: Int, x: Reader): Unit = {
    throw new SQLFeatureNotSupportedException("NClobs are not supported")
  }

  @throws[SQLException]
  override def setNClob(parameterIndex: Int, x: Reader, length: Long): Unit = {
    throw new SQLFeatureNotSupportedException("NClobs are not supported")
  }

  @throws[SQLException]
  override def setNString(parameterIndex: Int, x: String): Unit = {
    throw new SQLFeatureNotSupportedException("NStrings are not supported")
  }

  @throws[SQLException]
  override def setNull(parameterIndex: Int, sqlType: Int): Unit = {
    throw new SQLFeatureNotSupportedException("nulls are not supported")
  }

  @throws[SQLException]
  override def setNull(parameterIndex: Int, sqlType: Int, typeName: String): Unit = {
    throw new SQLFeatureNotSupportedException("nulls are not supported")
  }

  @throws[SQLException]
  override def setObject(parameterIndex: Int, x: AnyRef): Unit = {
    throw new SQLFeatureNotSupportedException("objects are not supported")
  }

  @throws[SQLException]
  override def setObject(parameterIndex: Int, x: Any, targetSqlType: Int): Unit = {
    throw new SQLFeatureNotSupportedException("objects are not supported")
  }

  @throws[SQLException]
  override def setObject(parameterIndex: Int, x: Any, targetSqlType: Int, scaleOrLength: Int): Unit = {
    throw new SQLFeatureNotSupportedException("objects are not supported")
  }

  @throws[SQLException]
  override def setRef(parameterIndex: Int, x: Ref): Unit = {
    throw new SQLFeatureNotSupportedException("refs are not supported")
  }

  @throws[SQLException]
  override def setRowId(parameterIndex: Int, x: RowId): Unit = {
    throw new SQLFeatureNotSupportedException("Row ids are not supported")
  }

  @throws[SQLException]
  override def setSQLXML(parameterIndex: Int, x: SQLXML): Unit = {
    throw new SQLFeatureNotSupportedException("SQLXMLs are not supported")
  }

  @throws[SQLException]
  override def setShort(parameterIndex: Int, x: Short): Unit = {
    setParameter(parameterIndex, NumericValue(BigDecimal(x)))
  }

  @throws[SQLException]
  override def setString(parameterIndex: Int, x: String): Unit = {
    setParameter(parameterIndex, StringValue(x))
  }

  @throws[SQLException]
  override def setTime(parameterIndex: Int, x: Time): Unit = {
    throw new SQLFeatureNotSupportedException("Time is not supported")
  }

  @throws[SQLException]
  override def setTime(parameterIndex: Int, x: Time, cal: Calendar): Unit = {
    throw new SQLFeatureNotSupportedException("Time is not supported")
  }

  @throws[SQLException]
  override def setTimestamp(parameterIndex: Int, x: Timestamp): Unit = {
    setParameter(parameterIndex, TimestampValue(x.getTime))
  }

  @throws[SQLException]
  override def setTimestamp(parameterIndex: Int, x: Timestamp, cal: Calendar): Unit = {
    throw new SQLFeatureNotSupportedException("Timestamp is not supported")
  }

  @throws[SQLException]
  override def setURL(parameterIndex: Int, x: URL): Unit = {
    throw new SQLFeatureNotSupportedException("URLs are not supported")
  }

  @throws[SQLException]
  override def setUnicodeStream(parameterIndex: Int, x: InputStream, length: Int): Unit = {
    throw new SQLFeatureNotSupportedException("Streams are not supported")
  }
}
